views/default/plugins/event_calendar/settings.php

```diff
@@ -1,5 +1,17 @@
 <?php
 
+if (event_calendar_is_upgrade_available()) {
+	echo '<div class="elgg-admin-notices">';
+	echo '<p>';
+	echo elgg_view('output/url', array(
+		'text' => elgg_echo('event_calendar:upgrade'),
+		'href' => 'action/event_calendar/upgrade',
+		'is_action' => true,
+	));
+	echo '</p>';
+	echo '</div>';
+}
```

Comment:
> Is there any way to execute the upgrade action directly without user input in the callback function within the 'upgrade', 'system' hook?

Sure, the handler could just run the upgrade immediately. There would be no need for a separate upgrade action. This might in fact be the best option.

> Also, the plugin specific upgrade might take a while (as in case with the converting of the annotations to relationships for the event_calendar plugin some sites might have many of them to process).

One option could be to wait for Elgg 2.0, which will allow also plugins to use the core upgrade system: https://github.com/Elgg/Elgg/pull/8217 But I guess that would be too far in the future?